example_test.go

```diff
@@ -85,6 +85,40 @@ func ExampleConnection_SelectTyped() {
 	// response is [{{} 1111 hello world}]
 }
 
+func ExampleFuture_GetIterator() {
+	var conn *tarantool.Connection
+	conn, err := example_connect()
+	if err != nil {
+		fmt.Printf("error in prepare is %v", err)
+		return
+	}
+	defer conn.Close()
+	const timeout = 3 * time.Second
```

Comment:
One of purposes of the push is to be able to send interactive signals while processing something long.

So instead of a function that tries to push everything rather quickly and return a result I've tried to use something slower:
```lua
local fiber = require('fiber')

local function push_func(x)
    for i = 1, x do
        box.session.push(i)
        fiber.sleep(0.3) -- 0.3 seconds
    end
    return x
end
```

I failed to run the example:
```bash
go clean -testcache && go test -run 'ExampleFuture_GetIterator' -timeout 100s
--- FAIL: ExampleFuture_GetIterator (0.50s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x4 pc=0x55e70b]

goroutine 1 [running]:
testing.(*InternalExample).processRunResult(0xc00013dae8, {0x0, 0x0}, 0xc0001cc018?, 0x0, {0x586840, 0x6dc270})
	/usr/local/go/src/testing/example.go:91 +0x4e5
testing.runExample.func2()
	/usr/local/go/src/testing/run_example.go:59 +0x11c
panic({0x586840, 0x6dc270})
	/usr/local/go/src/runtime/panic.go:838 +0x207
github.com/tarantool/go-tarantool_test.ExampleFuture_GetIterator()
	/home/georgymoiseev/Development/github/tarantool/go-tarantool/example_test.go:103 +0x1cb
testing.runExample({{0x5b653f, 0x19}, 0x5beef0, {0x5ba2e4, 0x24}, 0x0})
	/usr/local/go/src/testing/run_example.go:63 +0x28d
testing.runExamples(0xc00013dcd8, {0x6e10e0?, 0x4, 0x5?})
	/usr/local/go/src/testing/example.go:44 +0x186
testing.(*M).Run(0xc00011c460)
	/usr/local/go/src/testing/testing.go:1721 +0x689
github.com/tarantool/go-tarantool_test.runTestMain(0x1?)
	/home/georgymoiseev/Development/github/tarantool/go-tarantool/tarantool_test.go:1129 +0x1b5
github.com/tarantool/go-tarantool_test.TestMain(0x4413f1?)
	/home/georgymoiseev/Development/github/tarantool/go-tarantool/tarantool_test.go:1133 +0x19
main.main()
	_testmain.go:85 +0x1d3
exit status 2
FAIL	github.com/tarantool/go-tarantool	0.612s
```

I don't know if its some test timeout, but the only one documented `go test` timeout I've found is a `-timeout` flag which is set to be > 3s, so I suppose it's not the test framework fault.

Running iterator fetching cycle with 3s timeout and failing to process a 1.2s function is rather disappointing. It should either be fixed in the inside so it would be supported or described on the outside how one should use timeouts with long requests because for not it is not obvious (or not works).

Reply:
Thanks, I fixed the crash and a reason of it.

Indeed, this is an implementation limitation. It makes no sense to wait for a response longer than conn.opts.Timeout. It has [performance purposes](https://github.com/tarantool/go-tarantool/issues/48#issuecomment-334364328), so I don't change it. I have updated the comment for TimeoutResponseIterator.WithTimeout() with a note.

You can to test longer timeouts in a such way:
```
--- a/config.lua
+++ b/config.lua
@@ -56,10 +56,12 @@ local function simple_incr(a)
 end
 rawset(_G, 'simple_incr', simple_incr)
 
+fiber = require("fiber")
 local function push_func(cnt)
     for i = 1, cnt do
         box.session.push(i)
     end
+       fiber.sleep(1)
     return cnt
 end
 rawset(_G, 'push_func', push_func)
diff --git a/tarantool_test.go b/tarantool_test.go
index a9c0207..4243592 100644
--- a/tarantool_test.go
+++ b/tarantool_test.go
@@ -90,7 +90,7 @@ var spaceName = "test"
 var indexNo = uint32(0)
 var indexName = "primary"
 var opts = Opts{
-       Timeout: 500 * time.Millisecond,
+       Timeout: 5 * time.Second,
        User:    "test",
        Pass:    "test",
        //Concurrency: 32,

```

Reply:
I think it would be nice to go a bit further and log a warning when user calls `fut.GetIterator().WithTimeout(timeout)` with a timeout greater than request timeout. Request timeout is stored in `conn.opts.Timeout` and there are some log examples already: https://github.com/tarantool/go-tarantool/blob/a535b8e0224a63869dc911e8c39e72844929b322/connection.go#L65-L81

Reply:
> Shouldn't we reset future's timeout (and so move it to the tail of requests list due to the way timeouts implemented)?

It seems to me that this is logical. I have updated the peekFuture function:

```patch
--- a/connection.go
+++ b/connection.go
@@ -754,15 +773,34 @@ func (conn *Connection) putFuture(fut *Future, body func(*msgpack.Encoder) error
        }
 }
 
+func (conn *Connection) peekFuture(reqid uint32) (fut *Future) {
+       shard := &conn.shard[reqid&(conn.opts.Concurrency-1)]
+       pos := (reqid / conn.opts.Concurrency) & (requestsMap - 1)
+       shard.rmut.Lock()
+       defer shard.rmut.Unlock()
+
+       if conn.opts.Timeout > 0 {
+               fut = conn.getFutureImp(reqid, true)
+               pair := &shard.requests[pos]
+               *pair.last = fut
+               pair.last = &fut.next
+               fut.timeout = time.Now().Sub(epoch) + conn.opts.Timeout
+       } else {
+               fut = conn.getFutureImp(reqid, false)
+       }
+
+       return fut
+}
+
```

Reply:
Disregarding the approach we will implement here, we should document it in detail because it is definitely the place that may leave users confused.

Reply:
> Disregarding the approach we will implement here, we should document it in detail because it is definitely the place that may leave users confused.

I have added notes to Opts.Timeout and TimeoutResponseIterator.WithTimeout(). I think, the user will quickly notice that something is going wrong and look at the documentation.

Logging is also good. But it will be ugly in the current code. Future should know about Connection's params (Logger, Timeout)... It will be better to comment =)

Reply:
Ok, I don't have any more questions